hangups/ui/__main__.py

```diff
@@ -104,19 +114,51 @@ def __init__(self, refresh_token_path, keybindings, palette,
 
         self._urwid_loop.screen.set_terminal_properties(colors=palette_colors)
         self._urwid_loop.start()
+
+        coros = [self._connect(), self._coroutine_queue.consume()]
+
         # Enable bracketed paste mode after the terminal has been switched to
         # the alternate screen (after MainLoop.start() to work around bug
         # 729533 in VTE.
         with bracketed_paste_mode():
             try:
-                # Returns when the connection is closed.
-                loop.run_until_complete(self._client.connect())
+                # Run all the coros, until they all complete or one raises an
+                # exception. In the normal case, HangupsDisconnected will be
+                # raised.
+                loop.run_until_complete(asyncio.gather(*coros))
+            except HangupsDisconnected:
+                pass
             finally:
-                # Ensure urwid cleans up properly and doesn't wreck the
-                # terminal.
+                # Clean up urwid.
                 self._urwid_loop.stop()
+
+                # Cancel all of the coros, and wait for them to shut down.
+                task = asyncio.gather(*coros, return_exceptions=True)
+                task.cancel()
+                loop.run_until_complete(task)
```

Comment:
This would not wait for pending messages being sent.

Reply:
Is there a reason to wait rather than cancelling the task?

Reply:
The user might not be aware that message sending (possibly with an image upload) could take a few seconds.
> user: "I pressed enter, now I can exit hangups."


I know this behavior from other applications, for example email-applications alert the user for pending messages in the outbox when closing it.

Reply:
How about a progress indicator to show when there's a pending request?

Reply:
Good idea!
I reused the `LoadingWidget` with the addition of https://github.com/tdryer/hangups/pull/345/commits/e8395e6fc9067e600db2f54523c113f96542a8dd for #345 and then called it with `Disconnecting...`.